### PR TITLE
GpuShuffleCoalesceIterator acquire semaphore after host concat

### PR DESCRIPTION
**Pull request by abellina**
Closes https://github.com/NVIDIA/spark-rapids/issues/4395

This is a small optimization that was spotted while looking into NDS Q64 traces. With the change, Q64 can save up to 3 seconds (though it changes quite a bit from run to run). When executing this over all of NDS, it saved ~1 minute for the whole run chipping away at times a few hundred ms, up to 5 seconds for q94.

I saw some queries being slower, with the worst case being q42 (which for 1 sample out of 10 was 2x slower). I have not been able to reproduce this case, with all subsequent runs at 1x or above. This was a 3.8 second in the last weekly run, with the patch it's hovering between 3.6 and 4.5.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala [hunk 1]
```diff
@@ -151,9 +151,6 @@ class GpuShuffleCoalesceIterator(
   }
 
   private def concatenateBatch(): ColumnarBatch = {
-    // about to start using the GPU in this task
-    GpuSemaphore.acquireIfNecessary(TaskContext.get(), semWaitTime)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala:155]: The new code no longer acquires the semaphore when there's no columns as it did before.  Is that a change we really want to make here?  I can think of cases where we could end up exploding the row-count-only batch into GPU data without holding the semaphore.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala:155]: Oh right that's a great catch. We should add an unconditional acquire for the empty column case.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala:155]: Handled here: https://github.com/NVIDIA/spark-rapids/pull/4396/commits/bc2a56cc55789ddeb9c8b516d3506f0057b3a9ad. I ran a few pointed queries that saw benefits with this new commit (like q94), I don't see it negate the benefit of the original patch, so I think it's still a go.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Pull request merged**